pkg/build/strategies/sti/sti.go

```diff
@@ -596,9 +595,8 @@ func (builder *STI) Execute(command string, user string, config *api.Config) err
 	go dockerpkg.StreamContainerIO(errReader, &errOutput, func(a ...interface{}) { glog.Info(a...) })
 
 	err := builder.docker.RunContainer(opts)
-	if util.IsTimeoutError(err) {
-		// Cancel waiting for source input if the container timeouts
-		wg.Done()
+	if !config.LayeredBuild && !util.IsTimeoutError(err) {
+		defer wg.Wait()
```

Comment:
@gabemontero @bparees good job. Reading this method is an adventure... it deserves being rewritten, but not right now, let's get the issue fixed first.